readme.MD
# Magic the gathering pricer
## Description
Simple API that can look for MTG card prices through some websites.

Some sites are scrapped, others are fetched by their API.

Suggestions, fixes, issues or feature requests, feel free to fork and make a pull request or create an issue on this repository issue tracker.

## API documentation
Still in development

You can check for a card as
```
ENDPOINT/:from_where/card/:card_name
```

```:from_where``` can be
```
ligamagic
```
In the future I pretend to add some new websites.

```:card_name``` should be urlencoded (and spaced replaced by a + sign instead of %20)
```
Blood+Moon
```

E.g. ~~(This link doesn't really exists, just for explanation :) )~~
```
http://159.203.88.225:3000/ligamagic/card/Blood+Moon
```

The output will be something like that:
```
{
    status: "ok",
    data: {
        card: "Lua Sangrenta",
        currencies: [
            "BRL"
        ],
        sets: [
            "mma",
            "9e",
            "8e",
            "ch",
            "dk"
        ],
        prices: {
            mma: {
                BRL: [
                    "219.00",
                    "264.18",
                    "294.54"
                ]
            },
            9e: {
                BRL: [
                    "187.90",
                    "256.10",
                    "269.99"
                ]
            },
            8e: {
                BRL: [
                    "269.99",
                    "274.62",
                    "316.26"
                ]
            },
            ch: {
                BRL: [
                    "189.90",
                    "209.47",
                    "304.66"
                ]
            },
            dk: {
                BRL: [
                    "185.90",
                    "186.90",
                    "187.90"
                ]
            }
        },
        url: "http://ligamagic.com.br/?view=cards%2Fsearch&card=Blood+Moon"
    }
}
```

~~Notice that the price still not being fetched.... yet!~~

## How to run
To run this project is pretty simple:
- npm install
- node index.js


changelog.md
#Version 0.0.2
- Added to the json property prices with the prices (old sets property)
- Renamed the sets property to prices property
- Updated the way the prices are returned. Now it's an object with the
card set as key and returns another object with the currencies. In this object
you can access an array of prices (usually minor, average and major price).
- Added the currencies property that holds which currencies are being fetched
USD - US Dolar
BRL - Brazilian Real
You can find a list of abbreviations [here](http://www.xe.com/iso4217.php).


Example:
```javascript
{
    status: "ok",
    data: {
        card: "Lua Sangrenta",
        sets: ["mma"],
        currencies: ["BRL"],
        prices: {
            mma: {
                BRL: [
                    "219.00",
                    "264.18",
                    "294.54"
                ]
            }
        }
    }
}
```


index.js
var config = require('./config');
var proxy = require('./custom_modules/proxy');
var card_finder_factory = require('./custom_modules/card/finder/factory');
var request = require('request');
var cheerio = require('cheerio');
var http = require('http');
var express = require('express');
var app = new express();
var proxy_list = [];
var sites = ['ligamagic', 'tcgplayer', 'magicdomain'];

String.prototype.urldecode = function() {
  return decodeURIComponent(this).replace(/\+/g, ' ');
}

/* UPDATE the proxy_list */
var update_proxies = function (callback) {
  console.log("Updating proxy list");
  proxy.get(function(pl) {
    proxy_list = pl;
    console.log("Proxy list updated");
  });
  callback !== undefined && callback();

  setTimeout(update_proxies, 1000 * 60 * 60 * 24); //Update proxies every 24 hours
}
update_proxies();

app.get('/', function(req, res) {
  res.setHeader('Content-Type', 'application/json');
  res.json({status: 'error', data: { msg : 'Method not found' } });
});

app.get('/update_list', function(res, res) {
  res.setHeader('Content-Type', 'application/json');
  update_proxies(function() {
    res.json({status: 'ok', data: { msg : 'Proxy list updated' } });
  });
});

app.get('/get_proxy', function(res, res) {
  res.setHeader('Content-Type', 'application/json');
  if(proxy_list.length == 0) {
    update_proxies(function() {
      var random = parseInt(Math.random() * proxy_list.length, 10);
      res.json({status: 'ok', data: { proxy: proxy_list[random] } });
    });
  }else{
    var random = parseInt(Math.random() * proxy_list.length, 10);
    res.json({status: 'ok', data: { proxy: proxy_list[random] } });
  }
});

app.get('/:website/card/:cardname', function(req, res) {
  var website = req.params.website;
  var cardname = req.params.cardname;
  res.setHeader('Content-Type', 'application/json');

  if(sites.indexOf(website) == -1) {
    res.json({status: 'error', data: { msg : 'Source not found.' } });
  }else{
    try {
      var random = parseInt(Math.random() * proxy_list.length, 10);
      var finder = card_finder_factory.get(website);
      var find_promise = finder.find(cardname, proxy_list[random]);
      find_promise.then(function(card) {
        res.json({status: 'ok', data:
          {
            card: card.title.urldecode(),
            currencies : card.currencies,
            sets: card.sets,
            prices : card.prices,
            url: card.url
          }
        });
      }, function(err) {
        res.json({status: 'error', data:
          {
            msg: err
          }
        });
      });
    }catch(err) {
      console.log(err);
    }
  }
});

app.listen(3000);
console.log("Running on port 3000");


custom_modules/proxy.js
var request = require('request');
var cheerio = require('cheerio');
var Promise = require('promise');
var requestp = require('./requestp');
var proxy_list = [];

var request_options = {
  headers: {
    'User-Agent' : 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.155 Safari/537.36',
    'Accept' : 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8'
  }
};

//Simple prototype function
Date.prototype.yesterday = function() {
  var newDate = new Date(this.getTime());
  newDate.setDate(newDate.getDate() - 1);
  return newDate;
}

Date.prototype.toDBString = function() {
  return this.toISOString().slice(0, 10);
}

//Make the proxy request
var today = new Date();
var yesterday = today.yesterday();

module.exports = {
  get : function(get_callback) {
    request_options["url"] = "http://checkerproxy.net/getProxy?date=" + yesterday.toDBString();
    console.log("http://checkerproxy.net/getProxy?date=" + yesterday.toDBString());
    var request_yesterday = requestp(request_options).then(function(body) {
        var $ = cheerio.load(body);
        var total = $("ul").eq(1).find("li").length;
        if(total > 0) {
          $("ul").eq(1).find("li").each(function(i, elem) {
            proxy_list.push($(elem).text().replace(/^(\d+\.\d+\.\d+\.\d+:\d+)(\s+.*)?/gi, "$1"));
          });
        }
    }, function(err) {
      console.log("ERR getting yesterday proxy: " + err);
    });

    request_options["url"] = "http://checkerproxy.net/getProxy?date=" + today.toDBString();
    console.log("http://checkerproxy.net/getProxy?date=" + today.toDBString());
    var request_today = requestp(request_options).then(function(body) {
        var $ = cheerio.load(body);
        var total = $("ul").eq(1).find("li").length;
        if(total > 0) {
          $("ul").eq(1).find("li").each(function(i, elem) {
            proxy_list.push($(elem).text().replace(/^(\d+\.\d+\.\d+\.\d+:\d+)(\s+.*)?/gi, "$1"));
          });
        }
    }, function(err) {
      console.log("ERR getting today proxy: " + err);
    });

    return Promise.all([request_yesterday, request_today]).then(function() {
      /*
      var array_proxy = [];
      proxy_list.forEach(function(item) {
        array_proxy.push(item.split(":"));
      });
      */
      get_callback(proxy_list);
    });
  }
}


custom_modules/card/finder/ligamagic.js
var request = require('request');
var cheerio = require('cheerio');
var Promise = require('promise');
var requestp = require('../../requestp');

/* Resource for ligamagic */
module.exports = {
  find : function(cardname, proxy) {
    // Add some way to get the card with this resource
    console.log("Looking for " + cardname);
    console.log("Proxy: " + proxy);
    return new Promise(function(resolve, reject) {
      var request_options = {
        headers: {
          'User-Agent' : 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.155 Safari/537.36',
          'Accept' : 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8'
        },
        url: "http://ligamagic.com.br/?view=cards%2Fsearch&card=" + cardname
      };
      requestp(request_options)
        .then(function(body) {

          if(body.indexOf("0-0") !== -1)
            reject("Card not found");

          //read the data
          var $ = cheerio.load(body);
          var title = $(".cardTitle").text();
          var card_sets_obj = {};
          var card_set_array = [];
          var card = {};
          var moeda = "BRL";
          $(".overview td").each(function(i, elem) {
            var card_set = $("a", elem).attr("href");
            card_set = card_set.split("%3d")[1];
            card_sets_obj[card_set] = {}; //No prices found yet!
            card_sets_obj[card_set][moeda] = []; //Set prices as USD

            //@TODO:  Stop reading from the Javascript at the end of the page
            //        and scrappe the page instead. Dumbass! There's a table
            //        with the prices, lel!

            //Find the minor price for the card
            var regexMinorPrice = new RegExp("VETprecoMenor\\[" + i.toString() + "\\]\\s+=\\s+\"(.*)\";", "i");
            var minorPrice = body.match(regexMinorPrice)[1].replace(',', '.');
            card_sets_obj[card_set][moeda].push(minorPrice);

            //Find the medium price for the card
            var regexMediumPrice = new RegExp("VETprecoMedio\\[" + i.toString() + "\\]\\s+=\\s+\"(.*)\";", "i");
            var mediumPrice = body.match(regexMediumPrice)[1].replace(',', '.');
            card_sets_obj[card_set][moeda].push(mediumPrice);

            //Find the major price for the card
            var regexMajorPrice = new RegExp("VETprecoMaior\\[" + i.toString() + "\\]\\s+=\\s+\"(.*)\";", "i");
            var majorPrice = body.match(regexMajorPrice)[1].replace(',', '.');
            card_sets_obj[card_set][moeda].push(majorPrice);
          });

          card["title"] = title;
          card["prices"] = card_sets_obj;
          card["sets"] = Object.keys(card_sets_obj);
          card["currencies"] = [moeda];
          card["url"] = request_options["url"];

          resolve(card);
          //resolve("Misterious card info!");
        }, function(err) {
          reject(err); //Cascading promises
        });
    });
  }
}
